Bottom: 14c7bcfd24a1bb54eeb51dc3d806d2ddb4515d0d
Top:    f60f9e71e02f55dcdd1a0b5b59045d1ae7d96adc
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-12-09 23:23:26 +0000

Make HID->prompt_for and HID->fileselect return heap allocated memory

The callers should free this string free()


---

diff --git a/src/action.c b/src/action.c
index 07b2752..75d56ad 100644
--- a/src/action.c
+++ b/src/action.c
@@ -1408,8 +1408,6 @@ NotifyMode (void)
 		    Draw ();
 		  }
 		}
-
-	    /* free memory allocated by gui->prompt_for() */
 	    free (string);
 	  }
 	break;
@@ -4651,7 +4649,8 @@ ActionChangeName (int argc, char **argv, int x, int y)
 	case F_Layout:
 	  name =
 	    gui->prompt_for (_("Enter the layout name:"), EMPTY (PCB->Name));
-	  if (name && ChangeLayoutName (name))	/* XXX memory leak */
+	  /* NB: ChangeLayoutName takes ownership of the passed memory */
+	  if (name && ChangeLayoutName (name))
 	    SetChangedFlag (true);
 	  break;
 
@@ -4659,7 +4658,8 @@ ActionChangeName (int argc, char **argv, int x, int y)
 	case F_Layer:
 	  name = gui->prompt_for (_("Enter the layer name:"),
 				  EMPTY (CURRENT->Name));
-	  if (name && ChangeLayerName (CURRENT, name))	/* XXX memory leak */
+	  /* NB: ChangeLayerName takes ownership of the passed memory */
+	  if (name && ChangeLayerName (CURRENT, name))
 	    SetChangedFlag (true);
 	  break;
 	}
@@ -5951,7 +5951,8 @@ ActionNew (int argc, char **argv, int x, int y)
       CreateNewPCBPost (PCB, 1);
 
       /* setup the new name and reset some values to default */
-      PCB->Name = name;		/* XXX memory leak */
+      free (PCB->Name);
+      PCB->Name = name;
 
       ResetStackAndVisibility ();
       CreateDefaultFont ();
@@ -7622,6 +7623,8 @@ ActionImport (int argc, char **argv, int x, int y)
 	}
       else
 	AttributePut (PCB, "import::disperse", ds);
+      if (ARG (1) == NULL)
+        free (ds);
       return 0;
     }
 
diff --git a/src/change.c b/src/change.c
index df7f454..2545d84 100644
--- a/src/change.c
+++ b/src/change.c
@@ -1079,6 +1079,7 @@ ChangeTextName (LayerTypePtr Layer, TextTypePtr Text)
 bool
 ChangeLayoutName (char *Name)
 {
+  free (PCB->Name);
   PCB->Name = Name;
   hid_action ("PCBChanged");
   return (true);
@@ -1106,6 +1107,7 @@ ChangeElementSide (ElementTypePtr Element, LocationType yoff)
 bool
 ChangeLayerName (LayerTypePtr Layer, char *Name)
 {
+  free (CURRENT->Name);
   CURRENT->Name = Name;
   hid_action ("LayersChanged");
   return (true);
@@ -2280,7 +2282,7 @@ QueryInputAndChangeObjectName (int Type, void *Ptr1, void *Ptr2, void *Ptr3)
     }
   if (name)
     {
-      /* XXX Memory leak!! */
+      /* NB: ChangeObjectName takes ownership of the passed memory */
       char *old = ChangeObjectName (Type, Ptr1, Ptr2, Ptr3, name);
       if (old != (char *) -1)
 	{
diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index cfc1ed4..b7f9eea 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -302,9 +302,15 @@ nogui_prompt_for (const char *msg, const char *default_string)
     printf ("%s : ", msg);
 
   s = fgets (buf, 1024, stdin);
-  if (default_string && (s == NULL || buf[0] == 0))
-    strcpy (buf, default_string);
-  return buf;
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_string != NULL)
+        return strdup (default_string);
+      else
+        return strdup ("");
+    }
+  else
+    return strdup (buf);
 }
 
 /* FIXME - this could use some enhancement to actually use the other
@@ -322,10 +328,15 @@ nogui_fileselect (const char *title, const char *descr,
   else
     printf ("%s : ", title);
 
-  s = fgets (buf, 1024, stdin);
-  if (default_file && (s == NULL || buf[0] == 0))
-    strcpy (buf, default_file);
-  return buf;
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_file != NULL)
+        return strdup (default_file);
+      else
+        return NULL;
+    }
+  else
+    return strdup (buf);
 }
 
 static int
